packages/ra-core/src/core/CoreAdminRouter.tsx

```diff
@@ -1,107 +1,111 @@
 import React, {
     Children,
-    useEffect,
-    cloneElement,
-    createElement,
     ComponentType,
+    Fragment,
     ReactElement,
+    ReactNode,
+    useEffect,
 } from 'react';
-import { Route, Switch } from 'react-router-dom';
+import { Route, Routes } from 'react-router-dom';
 
-import RoutesWithLayout from './RoutesWithLayout';
+import { RoutesWithLayout } from './RoutesWithLayout';
 import { useLogout, useGetPermissions, useAuthState } from '../auth';
 import { useTimeout, useSafeSetState } from '../util';
 import { useScrollToTop } from './useScrollToTop';
 import {
     AdminChildren,
-    CustomRoutes,
     CatchAllComponent,
     LayoutComponent,
     LoadingComponent,
     CoreLayoutProps,
-    ResourceProps,
     RenderResourcesFunction,
-    ResourceElement,
+    ResourceDefinition,
+    ResourceProps,
 } from '../types';
+import { CustomRoutes, CustomRoutesProps } from './CustomRoutes';
+import { Resource } from './Resource';
+import { useRegisterResource } from './useRegisterResource';
 
-export interface AdminRouterProps extends CoreLayoutProps {
-    layout: LayoutComponent;
-    catchAll: CatchAllComponent;
-    children?: AdminChildren;
-    customRoutes?: CustomRoutes;
-    loading: LoadingComponent;
-    ready?: ComponentType;
-}
-
-type State = ResourceElement[];
-
-const CoreAdminRouter = (props: AdminRouterProps) => {
+export const CoreAdminRouter = (props: AdminRouterProps) => {
     const getPermissions = useGetPermissions();
     const doLogout = useLogout();
     const { authenticated } = useAuthState();
+    const registerResource = useRegisterResource();
     const oneSecondHasPassed = useTimeout(1000);
-    const [computedChildren, setComputedChildren] = useSafeSetState<State>([]);
+    const [resources, setResources] = useSafeSetState<ResourceDefinition[]>(
+        () => createResourcesFromChildren(props.children)
+    );
+    const [firstResource, setFirstResource] = useSafeSetState<
+        ResourceDefinition
+    >();
+    const [computedChildren, setComputedChildren] = useSafeSetState<
+        ReactNode
+    >();
+    const [hasFunctionChild, setHasFunctionChild] = useSafeSetState(false);
     useScrollToTop();
+
+    useEffect(() => {
+        setResources(createResourcesFromChildren(props.children));
+    }, [setResources, props.children]);
+
+    useEffect(() => {
+        resources.forEach(resource => {
+            registerResource(resource);
```

Comment:
if you rerender the Router, the same resource will be registered twice

Reply:
Unless registerResource takes care of it, which it does